src/components/organisms/EntityDetails/EntityDetailsContent/Settings/SettingsTests/SettingsTests.tsx

```diff
@@ -74,161 +68,170 @@ const SettingsTests: React.FC<{openDefinition(): void}> = ({openDefinition}) =>
-                ...firstItemInStep,
-                id,
-              };
-            }
+  const [isDelayModalVisible, setIsDelayModalVisible] = useState(false);
+  const [isTestModalVisible, setIsTestModalVisible] = useState(false);
+  const [currentGroup, setCurrentGroup] = useState<number | string>(0);
 
-            return {
-              ...firstItemInStep,
-              id,
-              type: testsData.find(item => item.name === firstItemInStep.test)?.type || '',
-              stopTestOnFailure: step.stopTestOnFailure,
-            };
-          })
-        : [],
-    [entityDetails?.steps, testsData]
-  );
+  const [steps, setSteps] = useState(initialSteps);
+  const [nodes, setNodes] = useState<ExtendedNode[]>([]);
+  const [edges, setEdges] = useState<Edge[]>([]);
 
-  const [currentSteps = initialSteps, setCurrentSteps] = useState<LocalStep[]>([]);
+  const onNodesChange = useCallback((changes: NodeChange[]) => setNodes(nds => applyNodeChanges(changes, nds)), []);
 
-  const wasTouched = currentSteps !== initialSteps;
+  const wasTouched = steps !== initialSteps;
 
   useEffect(() => {
-    if (currentSteps !== initialSteps) {
-      setCurrentSteps(initialSteps);
+    setSteps(initialSteps);
+  }, [testsData]);
+
+  const showModal = (value: string, group: number | string) => {
+    if (value === 'test') {
+      setIsTestModalVisible(true);
+    } else {
+      setIsDelayModalVisible(true);
     }
-  }, [initialSteps]);
 
-  const saveSteps = () => {
-    return updateTestSuite({
-      id: entityDetails.name,
-      data: {
-        ...entityDetails,
-        steps: currentSteps.map(step => {
-          return {
-            stopTestOnFailure: step.stopTestOnFailure,
-            execute: isV2
-              ? step.test
-                ? {name: step.test}
-                : {delay: step.delay}
-              : [step.test ? {test: step.test} : {delay: step.delay}],
-          };
-        }),
-      },
-    })
-      .then(res => displayDefaultNotificationFlow(res))
-      .then(() => {
-        notificationCall('passed', 'Steps were successfully updated.');
-      });
+    setCurrentGroup(group);
   };
 
-  const onSelectStep = (value: string) => {
-    if (value === 'delay') {
-      setIsDelayModalVisible(true);
+  const removeNode = (id: string, group: number) => {
+    if (steps[group].length === 1) {
+      setSteps([...steps.slice(0, group), ...steps.slice(group + 1)]);
     } else {
-      const {name, type} = JSON.parse(value);
-
-      setCurrentSteps([
-        ...currentSteps,
-        {
-          test: name,
-          id: nanoid(),
-          type,
-          stopTestOnFailure: false,
-        },
-      ]);
+      setSteps([...steps.slice(0, group), steps[group].filter(item => item.id !== id), ...steps.slice(group + 1)]);
     }
   };
 
-  const addDelay = (value: number) => {
-    setCurrentSteps([
-      ...currentSteps,
-      {
-        delay: `${value}ms`,
+  // add nodes
+  useEffect(() => {
```

Comment:
Probably it's worth to extract this `NodeExplorer` (or sth) component out of `SettingsTests`

Reply:
Done

Reply:
Sorry, I don't see it extracted, still there are things like:
```
          position: {x: index * 450, y: itemIndex * 100},
                  position: {x: index * 450 + 135, y: step.length * 100 - 20},
        position: {x: (index + 1) * 450 - 90, y: 11},
```

Reply:
I don't understand, you mean create functions for each of the positions? I doesn't make sense for me, would make it even more complicated

Reply:
It doesn't make anything complicated, it's making things simpler to understand without magic numbers. It can be solved with functions and constants.

Right now, the X/Y position here is magical - why there is `-90` or `-20`, when it's unrelated to anything? How to edit the spacing safely?

It could be achieved i.e. by something like:
```ts
// Configure
const addButtonWidth = 100;
const itemWidth = 300;
const itemVerticalSpace = 100;
const itemHorizontalSpace = 100;

// Setup helpers
const getIntersectionPosition = (group: number) => ({
  x: group * (itemWidth + itemVerticalSpace) + (itemVerticalSpace - addButtonWidth) / 2,
  y: 0,
});
const getAddPosition = (group: number) => ({ x: /*...*/, y: /*...*/ });
const getItemPosition = (group: number, index: number) => ({ x: /*...*/, y: /*...*/ });
```

Every library for visualizing data have such. As an example, [**d3**](https://d3js.org/) to solve that have some functions like:

```ts
const buildX = d3.scaleLinear().domain([0, 100]).range([0, 1000]);
const x = buildX(item.value);
```